### PR TITLE
bottom-most-compaction: use in test_gc_feedback + fix bugs

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

Adds manual compaction trigger; add gc compaction to test_gc_feedback

Part of https://github.com/neondatabase/neon/issues/8002

```
test_gc_feedback[debug-pg15].logical_size: 50 Mb
test_gc_feedback[debug-pg15].physical_size: 2269 Mb
test_gc_feedback[debug-pg15].physical/logical ratio: 44.5302 
test_gc_feedback[debug-pg15].max_total_num_of_deltas: 7 
test_gc_feedback[debug-pg15].max_num_of_deltas_above_image: 2 
test_gc_feedback[debug-pg15].logical_size_after_bottom_most_compaction: 50 Mb
test_gc_feedback[debug-pg15].physical_size_after_bottom_most_compaction: 287 Mb
test_gc_feedback[debug-pg15].physical/logical ratio after bottom_most_compaction: 5.6312 
test_gc_feedback[debug-pg15].max_total_num_of_deltas_after_bottom_most_compaction: 4 
test_gc_feedback[debug-pg15].max_num_of_deltas_above_image_after_bottom_most_compaction: 1
```

## Summary of changes

* Add the manual compaction trigger
* Use in test_gc_feedback
* Add a guard to avoid running it with retain_lsns
* Fix: Do `schedule_compaction_update` after compaction
* Fix: Supply deltas in the correct order to reconstruct value

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/timeline/compaction.rs [hunk 1]
```diff
@@ -959,15 +963,16 @@ impl Timeline {
...
     pub(crate) async fn compact_with_gc(
         self: &Arc<Self>,
         _cancel: &CancellationToken,
         ctx: &RequestContext,
     ) -> Result<(), CompactionError> {
+        use crate::tenant::storage_layer::ValueReconstructState;
         use std::collections::BTreeSet;
 
-        use crate::tenant::storage_layer::ValueReconstructState;
+        info!("running enhanced gc bottom-most compaction");
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:974]: add another line that logs something when we're done, using `scopeguard::defer!{}`.

File: pageserver/src/tenant/timeline/compaction.rs [hunk 2]
```diff
@@ -47,10 +47,14 @@ impl Timeline {
...
     pub(crate) async fn compact_legacy(
         self: &Arc<Self>,
-        _cancel: &CancellationToken,
+        cancel: &CancellationToken,
         flags: EnumSet<CompactFlags>,
         ctx: &RequestContext,
     ) -> Result<(), CompactionError> {
+        if flags.contains(CompactFlags::EnhancedGcBottomMostCompaction) {
+            return self.compact_with_gc(cancel, ctx).await;
+        }
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:56]: Hm, the flags aren't really meant for "modes".

I don't care too stronly about this for now, but, is there any functional reason why we don't treat bottommost compaction as another _step_ instead of returning early?

File: pageserver/src/tenant/timeline/compaction.rs [hunk 3]
```diff
@@ -1064,18 +1079,16 @@ impl Timeline {
...
+                            base_image = Some((*lsn, image.clone()));
+                            break;
                         }
                         crate::repository::Value::WalRecord(wal) => {
                             delta_above_base_image.push((*lsn, wal.clone()));
                         }
                     }
                 }
             }
-            delta_above_base_image.reverse();
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:1078]: This change seems like a correctness fix, orthogonal to what this PR is about.

I guess you found this because of the `test_gc_feedback`?
Is there a corresponding unit test?

File: pageserver/src/tenant/timeline/compaction.rs [hunk 4]
```diff
@@ -1208,6 +1226,9 @@ impl Timeline {
             let mut guard = self.layers.write().await;
             guard.finish_gc_compaction(&layer_selection, &compact_to, &self.metrics)
         };
+
+        self.remote_client
+            .schedule_compaction_update(&layer_selection, &compact_to)?;
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:1231]: This also seems like a drive-by bug fix. How did you discover it?

File: test_runner/performance/test_gc_feedback.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def test_gc_feedback(neon_env_builder: NeonEnvBuilder, zenbenchmark: NeonBenchma
             "checkpoint_distance": f"{1024 ** 2}",
             "compaction_target_size": f"{1024 ** 2}",
             # set PITR interval to be small, so we can do GC
-            "pitr_interval": "10 s",
+            "pitr_interval": "60 s",
```
Comment (problame) [inline test_runner/performance/test_gc_feedback.py:36]: Why this change?

**Review by problame — COMMENTED** [review]
better title for this PR would be:

```
bottom-most compaction: use in test_gc_feedback + fix bugs this revealed
```

And please enumerate / mention the list of bugs fixed in the PR description

File: pageserver/src/tenant/timeline/compaction.rs (hunk 2 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:56]: I feel for now it makes more sense to only run bottom-most compaction when the compaction API gets called, so that we can know the consequences of the compaction without interference from other parts of the system.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 3 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:1078]: Yes. It's tested as part of the benchmark, but I can also add a unit test for that based on NeonWalRecord.

File: test_runner/performance/test_gc_feedback.py (hunk 5 above)
Comment (skyzh) [reply test_runner/performance/test_gc_feedback.py:36]: If the interval is too small, the GC horizon would be above all existing delta+image layers, and the bottom-most compaction simply removes everything and creates image layers. By making the interval larger, we can have some deltas above the GC horizon, therefore covering the full code path in this benchmark.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 3 above)
Comment (problame) [reply pageserver/src/tenant/timeline/compaction.rs:1078]: Given that it's so essential, let's have that one. Can be done in a follow-up.

File: test_runner/performance/test_gc_feedback.py (hunk 5 above)
Comment (problame) [reply test_runner/performance/test_gc_feedback.py:36]: Hm, seems like an ugly source of non-determinism in this test.

Could we instead use `pitr_interval: 1y` and pass the right `gc_horizon` to `/compaction?legacy_enhanced...=yes`, so that the inputs are more deterministic?

Comment (skyzh) [reply test_runner/performance/test_gc_feedback.py:36]: because the benchmark runs for arbitrary time, it's also hard to pre-determine a gc_horizon for the benchmark in advance... anyways, I feel it's fine to have some non-determinism in benchmarks. The specific cases that gc_horizon cuts through some deltas, or above all deltas, are in Rust unit tests.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 4 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:1231]: By looking at the logs there are a bunch of warnings

Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:1231]: ...but I forgot the exact message

File: pageserver/src/tenant/timeline/compaction.rs (hunk 3 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:1078]: will do as a follow up

File: pageserver/src/tenant/timeline/compaction.rs (hunk 1 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:974]: resolved in https://github.com/neondatabase/neon/pull/8103/commits/4cc18325ab1c44ddc541475a0ec92a22668981a4

File: test_runner/performance/test_gc_feedback.py (hunk 5 above)
Comment (problame) [reply test_runner/performance/test_gc_feedback.py:36]: Fair, it's just that the benchmark results here will vary _massively_ depending on execution time => not really comparable.

But we don't have to solve this now.

**Review by problame — APPROVED** [review]
Please create a work issue for the `pitr_interval: ...` stuff and track it in your epic!
